src/iocore/net/BIO_fastopen.cc

```diff
@@ -167,18 +249,18 @@ fastopen_bread(BIO *bio, char *out, int outsz)
 static long
 fastopen_ctrl(BIO *bio, int cmd, long larg, void *ptr)
 {
+#if !defined(OPENSSL_IS_BORINGSSL) && (OPENSSL_VERSION_MAJOR >= 3)
   switch (cmd) {
-  case BIO_C_SET_CONNECT:
-    // We only support BIO_set_conn_address(), which sets a sockaddr.
-    ink_assert(larg == 2);
-    BIO_set_data(bio, ptr);
-    return 0;
+  case BIO_C_SET_FD:
+    BIO_set_shutdown(bio, int(larg));
```

Comment:
Shutdown??

Reply:
I had to add that code to avoid triggering this assert:  https://github.com/apache/trafficserver/blob/68a5ee56deee9c76c0204f046883f2752f1d7c1e/src/iocore/net/BIO_fastopen.cc#L100

Reply:
https://github.com/google/boringssl/blob/b34976cae99f8d1b864dbab31e20fc00d06acb09/crypto/bio/fd.c#L230

https://github.com/google/boringssl/blob/b34976cae99f8d1b864dbab31e20fc00d06acb09/crypto/bio/bio.c#L234

Reply:
Ok, I misunderstood what "shutdown" means here, and now setting it makes sense. But why only on OpenSSL 3+?

The odd thing is that it looks like `BIO_meth_get_ctrl` returns the ctrl function for Socket BIO, and it sets the shutdown flag. How does that not happen?

Reply:
I need to retest this.  Looking at the openssl code, it shouldn't be necessary.

Reply:
My testing was bogus, that change isn't necessary.